### PR TITLE
Use the new marginal sampler for latent gaussian models in the GP notebook

### PR DESCRIPTION
**Pull request by juanitorduz**
Closes https://github.com/blackjax-devs/blackjax/issues/238 

**Aim:** Use the data from https://blackjax-devs.github.io/blackjax/examples/GP_EllipticalSliceSampler.html to illustrate the implementation introduced in https://github.com/blackjax-devs/blackjax/pull/247 

For now, I have a running notebook that runs the example. I still wanted to push this branch and create a draft PR for feedback. What is missing (IMO):

**General**

- [x] Add text explaining the method and the results. I will do a very complete (verbose?) explanation for a newbie like me. We can trim the text to make it more succinct. 
- [x] Convert notebook to `md` (jupytext?)

**Details**

- [x] ~~I have used the implementation based on the unit-test of https://github.com/blackjax-devs/blackjax/pull/247 . At the moment I am setting the arbitrary value of `alpha=0.5` when defining the kernel. Should we use a calibration step as illustrated in https://github.com/blackjax-devs/blackjax/blob/main/examples/LogisticRegressionWithLatentGaussianSampler.md ?~~ I meant `delta` and not `alpha` 🤦 

Any feedback is welcome :)

### CODE REVIEW

**Comment by rlouf** [comment]
All sounds good! And yes, you will need to use jupytext to convert the notebook to `md`.

**Comment by AdrienCorenflos** [comment]
@juanitorduz sorry I didn't see this

> At the moment I am setting the arbitrary value of alpha=0.5 when defining the kernel. Should we use a calibration step as illustrated in https://github.com/blackjax-devs/blackjax/blob/main/examples/LogisticRegressionWithLatentGaussianSampler.md ?

alpha is a parameter in the prior, I am not sure what you mean by "calibration step". The calibration step in the notebook mentioned is for the delta parameter which corresponds (in a loose sense) to the step-size of MALA and other similar algos. Are you trying to learn alpha?

**Comment by juanitorduz** [comment]
Ahh! Thanks for the input! I was indeed confused! Now it makes sense. I will add the text and work on the notebook next week. 
Thanks for the input!

**Comment by AdrienCorenflos** [comment]
On a side note, good choice of example. The original paper considers GP classification and it shows that their method (mGrad, which is the one I implemented here) dominates the elliptical slice sampler (Ellipt):

image

File: examples/GP_Marginal.md [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+    A pure function that takes a rng key, a state and possibly some
+    parameters and returns a new state and some information about the
+    transition.
+```
+
+```python
+init, step = mgrad_gaussian(logprob_fn=logprob_fn, mean=jnp.zeros(n), covariance=Sigma)
+```
+
+For this particular sample, the only extra parameters is `delta`, which corresponds (in a loose sense) to the step-size of MALA and other similar algorithms.
```
Comment (rlouf) [inline examples/GP_Marginal.md:108]: Maybe set its value in `mgrad_gaussian`, it can be confusing otherwise to see mention of a parameter and not being able to find where it is defined.

File: examples/GP_Marginal.md [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+```
+
+## Run Sampler
+
+Now we proceed to run the sampler. First, we set the sampler parameters:
+
+```python
+# sampling parameters
+n_warm = 2000
+n_iter = 8000
```
Comment (rlouf) [inline examples/GP_Marginal.md:80]: I have tried `n_iter=1000` and it gives very good results. Since these notebooks run in CI we keep the number of sampling step as small as possible.

File: examples/GP_Marginal.md [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+
+```python
+n, d = 2000, 2
+length, scale = 1.0, 1.0
+y_sd = 1.0
+
+# fake data
+rng = jrnd.PRNGKey(10)
+kX, kf, ky = jrnd.split(rng, 3)
+X = jrnd.uniform(kX, shape=(n, d))
```
Comment (rlouf) [inline examples/GP_Marginal.md:50]:
```suggestion
kX, kf, ky = jrnd.split(rng, 3)

X = jrnd.uniform(kX, shape=(n, d))
```

Comment (rlouf) [reply examples/GP_Marginal.md:50]: Let's make this breathe a little!

Comment (rlouf) [reply examples/GP_Marginal.md:50]: Another way to make these breathe is to add comments that describe what the (main) quantities correspond to.

File: examples/GP_Marginal.md [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+%%time
+
+kernel = lambda key, x: step(key, x, 0.5)
+initial_state = init(f)
+
+states, info = inference_loop(jrnd.PRNGKey(0), init(f), kernel, n_warm + n_iter)
+samples = states.position[n_warm:]
+```
+
+## Evaluate Results
```
Comment (rlouf) [inline examples/GP_Marginal.md:139]:
```suggestion
## Diagnostics
```

File: examples/GP_Marginal.md [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+
+```python
+plt.figure(figsize=(8, 5))
+plt.hist(np.array(y), bins=50, density=True)
+plt.xlabel("y")
+plt.title("Histogram of data.")
+plt.show()
+```
+
+## Run Sampler
```
Comment (rlouf) [inline examples/GP_Marginal.md:73]:
```suggestion
## Sampling
```

File: examples/GP_Marginal.md [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+    language: python
+    name: python3
+---
+
+# Bayesian Regression With Latent Gaussian Sampler
+
+In this example, we want to illustrate how to use the marginal sampler implementation [`mgrad_gaussian`](https://blackjax-devs.github.io/blackjax/mcmc.html#blackjax.mgrad_gaussian) of the article [Auxiliary gradient-based sampling algorithms](https://rss.onlinelibrary.wiley.com/doi/abs/10.1111/rssb.12269). We do so by using the simulated data from the example [Gaussian Regression with the Elliptical Slice Sampler](https://blackjax-devs.github.io/blackjax/examples/GP_EllipticalSliceSampler.html). Please also refer to the complementary example [Bayesian Logistic Regression With Latent Gaussian Sampler](https://blackjax-devs.github.io/blackjax/examples/LogisticRegressionWithLatentGaussianSampler.html).
+
+
+## Prepare Notebook
```
Comment (rlouf) [inline examples/GP_Marginal.md:20]:
```suggestion
```

File: examples/GP_Marginal.md [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+import jax
+import jax.numpy as jnp
+import jax.random as jrnd
+import matplotlib.pyplot as plt
+import numpy as np
+
+from blackjax import mgrad_gaussian
+```
+
+## Generate Data
```
Comment (rlouf) [inline examples/GP_Marginal.md:32]:
```suggestion
```

File: examples/GP_Marginal.md [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+```python
+plt.figure(figsize=(8, 5))
+plt.hist(np.array(y), bins=50, density=True)
+plt.xlabel("y")
+plt.title("Histogram of data.")
+plt.show()
+```
+
+## Run Sampler
+
```
Comment (rlouf) [inline examples/GP_Marginal.md:74]: It would be nice to have a quick example of how the sampler works, like in the Elliptical Slice Sampling notebook.

**Review by rlouf — COMMENTED** [review]
It looks great! I modified the titles to match the notebook with the Elliptical Slice sampler (we may merge them eventually). It would also be nice to have a short explanation of how the sampler works if you feel like writing it.

File: examples/GP_Marginal.md (hunk 8 above)
Comment (juanitorduz) [reply examples/GP_Marginal.md:74]: Yeah! I want to do this... unfortunately I have not been able to access the paper (I am trying through old access from my university). Is there a pre-print available?

File: examples/GP_Marginal.md (hunk 2 above)
Comment (juanitorduz) [reply examples/GP_Marginal.md]: reduced in https://github.com/blackjax-devs/blackjax/pull/398/commits/a3a4253d7db1cff2a4b8d9baef97100eb37e84eb

File: examples/GP_Marginal.md (hunk 1 above)
Comment (juanitorduz) [reply examples/GP_Marginal.md]: I added it as a `kwarg` in https://github.com/blackjax-devs/blackjax/pull/398/commits/a3a4253d7db1cff2a4b8d9baef97100eb37e84eb

File: examples/GP_Marginal.md (hunk 8 above)
Comment (rlouf) [reply examples/GP_Marginal.md:73]: Yes! https://arxiv.org/abs/1610.09641

Comment (AdrienCorenflos) [reply examples/GP_Marginal.md:73]: https://arxiv.org/abs/1610.09641

Comment (juanitorduz) [reply examples/GP_Marginal.md:73]: OMG 🤦 I can not believe I did not find this! I think I was keen to get the published version. Anyway, let me give a read and try to write this section myself, and then I'll collect feedback. Thanks!

Comment (AdrienCorenflos) [reply examples/GP_Marginal.md:73]: Btw, this is the "executive summary" of the method I have in the paper I am currently writing.
image

**Comment by juanitorduz** [comment]
@rlouf @AdrienCorenflos In https://github.com/blackjax-devs/blackjax/pull/398/commits/687f2950522c4379c4829191a03e841f7a83442b I tried to express in my words how I understand the sampler works. I have a math background, which is far from this area so I feel it is at the moment out of my reach to have a very detailed and transparent description of the details behind the sampler (if you want to talk about geometric analysis I feel a bit better prepared hehe). Hence, feel free to modify / add or erase parts that you believe are confusing or incorrect.

**Comment by AdrienCorenflos** [comment]
> @rlouf @AdrienCorenflos In [687f295](https://github.com/blackjax-devs/blackjax/commit/687f2950522c4379c4829191a03e841f7a83442b) I tried to express in my words how I understand the sampler works. I have a math background, which is far from this area so I feel it is at the moment out of my reach to have a very detailed and transparent description of the details behind the sampler (if you want to talk about geometric analysis I feel a bit better prepared hehe). Hence, feel free to modify / add or erase parts that you believe are confusing or incorrect.

I made some changes to the explanation of the algo. 
https://github.com/AdrienCorenflos/blackjax/tree/patch-2
@rlouf if you could check and merge?

**Comment by rlouf** [comment]
I'll take a look early next week, thanks.

**Comment by rlouf** [comment]
I added the changes suggested by @AdrienCorenflos (and added him as co-author) and rebased the commits. It looks good to me, can you take a look one last time before I merge?

**Review by rlouf — APPROVED** [review]

**Comment by juanitorduz** [comment]
Thank you for the feedback and corrections @AdrienCorenflos ! LGTM.

**Comment by rlouf** [comment]
Merging now, great work everyone!
